content/post/performance-metrics-histogram.md

```diff
@@ -0,0 +1,245 @@
+
+You might be thinking about other places where similar conclusions come up. Theo Schlossnagle's excellent posts at circonus.com are excellent examples (see [The Uphill Battle for Visibility](https://www.circonus.com/2016/07/uphill-battle-visibility/), [The Problem with Math: Why Your Monitoring Solution is Wrong](https://www.circonus.com/2015/02/problem-math/), and [Percentages Aren’t People](https://www.circonus.com/2016/06/percentages-arent-people/)). 
+
+You may also be thinking, "Gee, this sounds a lot like an [HdrHistogram](http://hdrhistogram.github.io/HdrHistogram/)!" Have you looked into using an HdrHistogram? And indeed, we have.
+But in the end, we prefer a different approach for several reasons: 
+
+1. While HdrHistograms are amazing, they are more expensive in both time and space than a basic histogram with a small number of fixed bins.
+2. While the added resolution is handy if one needs to answer arbitrary percentile questions, this isn't currently the case for Wallaroo. The current situation requires answers to very explicit questions, such as _"is the 99th percentile latency below 1ms?"_
+3. There is currently no HdrHistogram implementation for Pony. (Note that this is no longer the case, as there is an excellent port by [Darach Ennis](https://github.com/darach/hdr_histogram_pony))
+4. The cost or recording an event _is still too high_ for very high throughput loads.
+
+## Powers of 2 Histogram
+
+The binary representation of numbers offers a very useful optimization: one can obtain the nearest power of 2 that is greater than a current number by counting the leading zeros in a fixed width integer representation of that number. More importantly, nearly all modern CPUs have hardware support for this operation, which we can leverage.
+
+For example, if the 32-bit big-endian representation of the number `100,000` is `00000000000000011000011010100000`, which has 15 leading zeros, then its nearest power of 2 which is greater or equal to it can be calculated as   
+2^(32-15) = 2^17 = 131072.  
+This comes in handy since we can use a single instruction to count the leading zeros of an unsigned integer.
+
+## Implementation
+
+To achieve the lowest possible performance measurement overhead, we chose to use the powers-of-2 histogram as the internal metrics counter in each Wallaroo worker. The way this works is by taking a nanosecond timestamp immediately before and after the points of interest (e.g. before and after a computation, or at the input and output boundaries of a worker, or at the source and sink of a pipeline). The pair of timestamps is converted into a nanosecond delta, which is then recorded in the appropriate histogram for the particular step, worker, or pipeline. The histogram is maintained over a period of 1 second, after which a new histogram is created and the old one is sent out to an external metrics service (such as Wallaroo's bundled [Metrics UI](https://docs.wallaroolabs.com/book/metrics/metrics-ui.html)).
```

Comment:
Here we say a `histogram is maintained over a period of 1 second` but earlier we say `a histogram is produced per step per metrics period (typically 2 seconds)`. These numbers should likely be consistent.